src/details/ArborX_DetailsSortUtils.hpp

```diff
@@ -65,76 +65,136 @@ namespace ArborX
+  // space as argument
+  Kokkos::BinSort<ValuesType, CompType, typename ValuesType::device_type,
+                  SizeType>
+      bin_sort(values, CompType(n / 2, result.min_val, result.max_val), true);
+  bin_sort.create_permute_vector();
+  bin_sort.sort(values);
+}
+
+#if defined(KOKKOS_ENABLE_CUDA) ||                                             \
+    (defined(KOKKOS_ENABLE_HIP) && defined(ARBORX_ENABLE_ROCTHRUST))
+template <typename ValuesType>
+void sort(
+#if defined(KOKKOS_ENABLE_CUDA)
+    Kokkos::Cuda const &space,
+#else
+    Kokkos::Experimental::HIP const &space,
+#endif
+    ValuesType &values)
+{
+  static_assert(Kokkos::is_view<ValuesType>::value, "");
+  static_assert(std::is_same<std::decay_t<decltype(space)>,
+                             typename ValuesType::execution_space>::value,
+                "");
+
+  int const n = values.extent(0);
+
+#if defined(KOKKOS_ENABLE_CUDA)
+  auto const execution_policy = thrust::cuda::par.on(space.cuda_stream());
+#else
+  auto const execution_policy = thrust::hip::par.on(space.hip_stream());
+#endif
+
+  using ValueType = typename ValuesType::value_type;
+
+  auto values_first = thrust::device_ptr<ValueType>(values.data());
+  auto values_last = thrust::device_ptr<ValueType>(values.data() + n);
+  thrust::sort(execution_policy, values_first, values_last);
+}
+#endif
+
+template <typename ExecutionSpace, typename KeysType, typename ValuesType>
+void sortByKey(ExecutionSpace const &space, KeysType &keys, ValuesType &values)
+{
+  static_assert(Kokkos::is_view<KeysType>::value, "");
+  static_assert(Kokkos::is_view<ValuesType>::value, "");
 
-  Kokkos::BinSort<ViewType, CompType, typename ViewType::device_type, SizeType>
-      bin_sort(view, CompType(n / 2, result.min_val, result.max_val), true);
+  int const n = keys.extent(0);
+
+  using SizeType = unsigned int;
+  using ValueType = typename KeysType::value_type;
+  using CompType = Kokkos::BinOp1D<KeysType>;
+
+  Kokkos::MinMaxScalar<ValueType> result;
+  Kokkos::MinMax<ValueType> reducer(result);
+  parallel_reduce("ArborX::Sorting::find_min_max_view",
+                  Kokkos::RangePolicy<ExecutionSpace>(space, 0, n),
+                  Kokkos::Impl::min_max_functor<KeysType>(keys), reducer);
+  if (result.min_val == result.max_val)
+    return;
+
+  Kokkos::BinSort<KeysType, CompType, typename KeysType::device_type, SizeType>
+      bin_sort(keys, CompType(n / 2, result.min_val, result.max_val), true);
   bin_sort.create_permute_vector();
-  bin_sort.sort(view);
+  bin_sort.sort(keys);
   // FIXME Kokkos::BinSort is currently missing overloads that an execution
   // space as argument
 
-  return bin_sort.get_permute_vector();
+  auto permute = bin_sort.get_permute_vector();
+  auto values_clone = clone(values);
+  Kokkos::parallel_for(
+      "ArborX::Sorting::apply_permutation",
+      Kokkos::RangePolicy<ExecutionSpace>(space, 0, n),
+      KOKKOS_LAMBDA(int const i) { values(i) = values_clone(permute(i)); });
```

Comment:
??
```C++
bin_sort.sort(values);

Reply:
A fair point. Did not think of that because `sort()` name invokes heavy computation, while in Kokkos `create_permute_vector()` is the heavy one.